### PR TITLE
Add location filter to careers pages

### LINKED ISSUE #189: "Sort by locations" doesn't sort by locations...
The new Canonical careers website makes it very hard to look for open positions in a given country/geographical area.

# Steps to reproduce

1. Go to https://canonical.com/careers/all
2. For "Sort by", selection "Location"

# Expected results

Jobs are gathered by countries or at least geographical locations

# Actual results

As of now, here are the first 10 entries:

- "Americas"
- "Americas"
- "Americas"
- "Americas or European timezone"
- "Americas, EMEA"
- "Anywhere"
- "EMEA"
- "EMEA"
- "EMEA"
- "EU - Eastern US"

It's clearly only sorted alphabetically.

# Better approach

A better solution would involve having one or more proper location tag(s) for all the job descriptions (instead of a simple string as it's apparently currently the case), including a "Remote" one, and allowing users to filter by these. In this way, a candidate could filter for positions in the Taipei office easily (by selection "Asia" or "Taiwan"), or for remote positions in Europe (by ticking "Remote" and "Europe" for instance).

### PR DESCRIPTION
**Pull request by MeltySnow**
## Done

Added the ability to filter jobs by location on the careers pages

## QA

- Check out this feature branch
- Run the site using the command `./run serve`
- View the site locally in your web browser at: http://0.0.0.0:8002/careers/all
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)
- Ensure that jobs are properly filtered by location, and that the new field is compatible with the existing "Filter by" and "Sort by" fields.
- Ensure that the code is sensible and in line with any Canonical style guides

## Issue / Card

Fixes #189

### CODE REVIEW

File: static/js/careers-filter-and-sort.js [hunk 1]
```diff
@@ -5,42 +5,168 @@
...
+          break;
+        }
+      }
+
+      if(locationsList.length > 0)
+      {
+        locationsList = locationsList.slice(0, locationsList.length-1);
+      }
+      jobsList[n].setAttribute("location-filter", locationsList);
+    }
```
Comment (sowasred2012) [inline static/js/careers-filter-and-sort.js:83]: It's really good, works well - I had a suggestion that would reduce some repetition, see what you think:
```suggestion
const regions = {
        europe: ["emea", "slovakia", "bratislava", "europe", "uk", "germany", "berlin", "london", "worldwide"],
        americas: [ "americas", "southwest", "san francisco", "usa", "austin", "texas", "tx", "brazil", "seattle", "america", "worldwide"],
        asia: ["apac", "taiwan", "taipei", "beijing", "china", "worldwide"],
        "middle-east": ["emea", "worldwide"],
        africa: ["emea", "worldwide"],
        oceania: ["apac", "worldwide"],
      };

      const jobsList = document.querySelector(".js-job-list").children;

      for (var n = 0; n < jobsList.length; n++) {
        const location = jobsList[n].getAttribute("data-location");
        var locationsList = "";

        for (var region in regions) {
          var regionalLocations = regions[region];

          for (var i = 0; i < regionalLocations.length; i++) {
            if (location.toLowerCase().includes(regionalLocations[i])) {
              locationsList += region + " ";
            }
          }
        }

        if (locationsList.length > 0) {
          locationsList = locationsList.slice(0, locationsList.length - 1);
        }
        jobsList[n].setAttribute("location-filter", locationsList);
      }
```

**Comment by MeltySnow** [comment]
Thanks for the feedback; I've pushed up changes that should hopefully address @anthonydillon's formatting issues, and incorporated @sowasred2012's suggestion to use nested for loops in the parseLocations() function.

File: static/js/careers-filter-and-sort.js [hunk 2]
```diff
@@ -75,16 +155,38 @@
...
+        if (node.getAttribute("location-filter").includes(filterBy.location)) {
+          if (node.classList.contains("u-hide")) {
+            node.classList.remove("u-hide");
+          }
+          } else {
+            if (!node.classList.contains("u-hide")) {
+              node.classList.add("u-hide");
+          }
+          numberOfJobsDisplayed--;
+        }
```
Comment (sowasred2012) [inline static/js/careers-filter-and-sort.js:184]: Small indentation tweak here - at first glance this looked like a verbose way of writing `node.classList.toggle("u-hide")`
```suggestion
        } else {
          if (!node.classList.contains("u-hide")) {
            node.classList.add("u-hide");
          }
          numberOfJobsDisplayed--;
        }
      } else if (filterBy.filterValue === "all" && filterBy.location !== "all") {
        if (node.getAttribute("location-filter").includes(filterBy.location)) {
          if (node.classList.contains("u-hide")) {
            node.classList.remove("u-hide");
          }
        } else {
          if (!node.classList.contains("u-hide")) {
            node.classList.add("u-hide");
          }
          numberOfJobsDisplayed--;
        }
```

**Comment by MeltySnow** [comment]
Done!

**Pull request merged**